plutus/budget_manager/gcp_helper.py

```diff
@@ -0,0 +1,282 @@
+            self.logger.error(f"Request failed due to GoogleAPICallError: {err}")
+            metrics.incr("gcp_api_error_count", tags = ["type:billing.update.apicall", f"display_name:{changed_budget['display_name']}"])
+        except RetryError as err:
+            self.logger.error(f"Request failed due to retryable error and retry attempts failed: {err}")
+            metrics.incr("gcp_api_error_count", tags = ["type:billing.update.retry", f"display_name:{changed_budget['display_name']}"])
+        except ValueError as err:
+            self.logger.error(f"Requst failed likely due to invalid parameters. ValueError: {err}")
+            metrics.incr("gcp_api_error_count", tags = ["type:billing.update.value", f"display_name:{changed_budget['display_name']}"])
+
+
+    def get_and_update_or_create_budget(self, project):
+        '''
+        This method takes a ProjectBudget object constructed from the yaml config and does one of two things:
+        1. Gets the budget, compares values with GCP and updates budget if changes have occurred
+        2. Creates a budget in GCP
+
+        The method will return the budget object or None if error
+        '''
+
+        project_id = project.project_id
+        project_number = self.get_project_number(project_id)
+
+        budgets = self.get_budgets_by_project(project, project_number)
+
+        if len(budgets) == 0:
+            budget = create_budget(project)
+            return budget
+        elif len(budgets) >= 1:
+            for budget in budgets:
+                if budget.display_name == project.display_name:
+                    # We have found the correct budget to compare, since we programatically create display names
+                    budget_dict = protobuf_to_dict(budget, use_enum_labels=True)
+                    same, changed_budget = self.sync_config_with_gcp_budget(project, budget_dict, project_number)
+
+                    if not same and changed_budget is not None:
+                        self.logger.info(f"Changes detected. Updating budget to {changed_budget}")
+                        new_budget = self.update_budget(changed_budget)
+                        return new_budget
+                    elif changed_budget is None:
+                        self.logger.error("Error encountered while comparing config and gcp budget. Exiting...")
+                        sys.exit(1)
+                    else:
+                        self.logger.info(f"No changes detected with budget {budget_dict['display_name']}")
+                        return budget
+
+            # No budgets for project matched the config, create new budget
+            self.logger.info(f"No existing plutus budget found for {project.display_name}, creating...")
+            new_budget = self.create_budget(project)
+            return new_budget
+        else:
+            self.logger.error(f"Unexpected error for: {project.project_id}. Length of matching projects found was undefined.")
```

Comment:
Would this be a scenario when the config is incorrect?

Reply:
Actually I don't think this will ever occur. removing

Reply:
good question. I don't expect this case to ever hit because any misconfigured projectid should throw an exception earlier. I just went over use cases and I dont think this should ever hit, should i rm this ?

Reply:
If the code path is not going to be hit might as well remove it.